### ISSUE TITLE
Updating a route requires path, host, method to be set

### ISSUE THREAD

**Issue by augwa** [Kong/kong#3361]
### Summary

According to the documentation, it stats that path, host, method are conditionally required as such that one of the 3 must be set. However all 3 are required when doing an update.

### Steps To Reproduce

1. Create a Service
> request
```
curl -X POST \
  http://172.30.0.102:8001/services \
  -H 'cache-control: no-cache' \
  -H 'content-type: application/x-www-form-urlencoded' \
  -H 'postman-token: b56e1fd6-c4c9-1e3f-50e4-533e3249e982' \
  -d 'name=bug-report&protocol=https&host=127.0.0.1&path=%2F&port=443'
```
> response
```
{
    "host": "127.0.0.1",
    "created_at": 1522350788,
    "connect_timeout": 60000,
    "id": "be8493de-27f4-49c1-9606-cc3e24e8a7e4",
    "protocol": "https",
    "name": "bug-report",
    "read_timeout": 60000,
    "port": 443,
    "path": "/",
    "updated_at": 1522350788,
    "retries": 5,
    "write_timeout": 60000
}
```
2. Create a Route
> request
```
curl -X POST \
  http://172.30.0.102:8001/routes \
  -H 'cache-control: no-cache' \
  -H 'content-type: application/x-www-form-urlencoded' \
  -H 'postman-token: ffbdedb3-b935-7774-938c-70b8fa634239' \
  -d 'protocols%5B%5D=http&protocols%5B%5D=https&paths%5B%5D=%2Ftest&service.id=be8493de-27f4-49c1-9606-cc3e24e8a7e4'
```
> response
```
{
    "created_at": 1522350912,
    "strip_path": true,
    "hosts": null,
    "preserve_host": false,
    "regex_priority": 0,
    "updated_at": 1522350912,
    "paths": [
        "/test"
    ],
    "service": {
        "id": "be8493de-27f4-49c1-9606-cc3e24e8a7e4"
    },
    "methods": null,
    "protocols": [
        "http",
        "https"
    ],
    "id": "1e61ab66-8ce7-4036-85a7-cdd0089ea3c7"
}
```
3. Update route by supplying only a path, host or method, example shown with method but all 3 produce similar results.
> request
```
curl -X PATCH \
  http://172.30.0.102:8001/routes/1e61ab66-8ce7-4036-85a7-cdd0089ea3c7 \
  -H 'cache-control: no-cache' \
  -H 'content-type: application/x-www-form-urlencoded' \
  -H 'postman-token: 36e12a0e-6dfe-e9d0-4dd5-db74d72f2b64' \
  -d methods%5B%5D=POST
```
> response
```
{
    "fields": {
        "hosts": "field required for entity check",
        "paths": "field required for entity check"
    },
    "name": "schema violation",
    "code": 2,
    "message": "2 schema violations (hosts: field required for entity check; paths: field required for entity check)"
}
```

### Additional Details & Logs

- Kong version 0.13.0

**Comment by hishamhm** [comment]
Hi @augwa -- yes, this is working as intended. It's a bit annoying, but when performing a PATCH you need to pass all three explicitly, even if the other two are `null`.

We need all three values when doing a PATCH so we can validate that at least one of them is set. Otherwise, in a route that had `"hosts": ["example.com"], "paths": null, "methods": null`, giving it a partial update with just `"hosts": null` would be able to break the constraint that one of the three must be set.

You may ask, "couldn't Kong just fetch the data first, apply the update to the object, check the constraint, and then update the DB only if everything is okay"? This would cause a concurrency problem: what if someone else makes an edit it between the data is fetched for checking and the update is committed? That update would be lost.

You may then ask, "well, you could lock this with a transaction then!"... but Kong supports both Postgres (which has transactions) and Cassandra (which does not). 

So, to ensure DB consistency and identical behavior with both backend DBs, we had to make PATCH demand all three values.

Perhaps the error message can be improved to something like `"field must be explicitly given due to validity check (use null if value should be unset)"`. Would that had made things less confusing?

**Comment by augwa** [comment]
Actually, just thinking about it. Why would Kong have to fetch the values at all? The only time this becomes an issue is if you want to set one of them as null, without passing anything else. In this case it should perhaps require passing in another one in as non-empty value, because as you mentioned there could be a concurrency issue where another request comes in and set another as null and now all of them all null. But in the case where I'm giving a value to any of those 3 that isn't null, then it shouldn't matter if one or all 3 is passed in, because it will always pass validation if one of them contains a value.

**Comment by hishamhm** [comment]
Your reasoning makes total sense. One current limitation to implement it is that the schema validation rules currently only look at field names which ones are needed to perform a constraint check, and not their values (so it can't special-case based on the knowledge that the new input value is non-null). This needs to be improved in the schema library.

**Comment by hishamhm** [comment]
@augwa I just pushed PR #3364 to relax the schema validation rules.

**Comment by thibaultcha** [comment]
Just merged ##3364! Thus closing this, it will be included in the next major release. Thanks for the suggestion @augwa!
